docs/docs/dev_docs/contracts/portals/main.md

```diff
@@ -0,0 +1,331 @@
+
+    // Transfer the tokens to the user
+    underlying.transfer(_recipient, _amount);
+
+    return entryKey;
+  }
+```
+
+
+## How to deploy a contract with a portal
+- Deploy L1 
+- Deploy l2
+- Initialize l1 with l2 address.
+
+## Standards
+
+### Structure of messages
+The application developer should consider creating messages that follow a structure like function calls, e.g., using a function signature and arguments. This will make it easier to ensure that they are not making messages that could be misinterpreted by the recipient. 
+
+An example of a bad format would be using `amount, token_address, recipient_address` as the message for a withdraw function and `amount, token_address, on_behalf_of_address` for a deposit function. Any deposit could then also be mapped to a withdraw or vice versa. 
+
+```solidity 
+// Don't to this!
+bytes memory message = abi.encode(
+  _amount, 
+  _token, 
+  _to
+);
+
+// Do this!
+bytes memory message abi.encodeWithSignature(
+  "withdraw(uint256,address,address)", 
+  _amount, 
+  _token, 
+  _to
+);
+```
+
+### Error Handling
+
+Handling error when moving cross chain can quickly get tricky. Since the L1 and L2 calls are practically async and independent of each other, the L1 part of a deposit might execute just fine, with the L2 part failing. If not handling this well, the funds might be lost forever! The contract builder should therefore consider in what ways his application can fail cross chain, and handle those cases explicitly.
+
+First, entries in the outboxes **SHOULD** only be consumed if the execution is successful. For an L2 -> L1 call, the L1 execution can revert the transaction completely if anything fails. As the tx can be atomic, the failure also reverts the consumption of the entry.
+
+If it is possible to enter a state where the second part of the execution fails forever, the application builder should consider including additional failure mechanisms (for token withdraws this could be depositing them again etc).
+
+Generally it is good practice to keep cross-chain calls simple to avoid too many edge cases and state reversions.
+
+:::info
+Error handling for cross chain messages is handled by the application contract and not the protocol. The protocol only delivers the messages, it does not ensure that they are executed successfully.
+:::
+
+### Cancellations
+
+A special type of errors is underpriced transactions, here meaning that message is inserted on L1, but the fee available to the sequencer is too low to be included in a rollup block.
+
+For the case of token bridges, this could lead to funds being locked in the bridge forever, as the funds are locked but the message never arrives on L2 to mint the tokens. To address this, the `Inbox` supports cancelling messages after a deadline. However, this must be called by the portal itself, as it will need to "undo" the state changes is made (for example by sending the tokens back to the user).
```

Comment:
```suggestion
For the case of token bridges, this could lead to funds being locked in the bridge forever, as funds are locked but the message never arrives on L2 to mint the tokens. To address this, the `Inbox` supports cancelling messages after a deadline. However, this must be called by the portal itself, as it will need to "undo" the state changes is made (for example by sending the tokens back to the user).
```